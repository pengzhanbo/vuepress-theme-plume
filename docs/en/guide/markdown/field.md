---
title: Field Container
icon: solar:text-field-linear
createTime: 2025/10/08 09:55:17
permalink: /en/guide/markdown/field/
badge: New
---

## Overview

In Markdown, use the `::: field` container to describe field information, including field name,
field type, whether it's required, default value, description, and other details.

It's suitable for describing fields in configurations, component Props, and similar scenarios.

You can also use the additional `:::: field-group` container to group multiple `::: field` containers.

## Enable

This feature is disabled by default. You need to enable it in the `theme` configuration.

```ts title=".vuepress/config.ts"
export default defineUserConfig({
  theme: plumeTheme({
    markdown: {
      field: true, // [!code ++]
    },
  })
})
```

## Syntax

```md
<!-- Single Field -->
::: field name="Field Name" type="Type" required default="Default Value"
Field description information
:::

<!-- Field Group -->
:::: field-group

::: field name="Field Name" type="Type" required default="Default Value"
Field description information
:::

::: field name="Field Name" type="Type" required default="Default Value"
Field description information
:::

::::
```

## Properties

:::: field-group
::: field name="name" required type="string"
Field name
:::

::: field name="type" type="string" optional
Field type
:::

::: field name="required" type="boolean" optional
Whether the field is required
:::

::: field name="optional" type="boolean" optional
Whether the field is optional
:::

::: field name="deprecated" type="boolean" optional
Whether the field is deprecated
:::

::: field name="default" type="string" optional
Default value. If it's an empty string, use `default="''"`
:::
::::

## Examples

**Input:**

```md
::: field name="theme" type="ThemeConfig" required default="{}"
Theme configuration
:::

::: field name="enabled" type="boolean" optional default="true"
Whether enabled
:::
```

**Output:**

::: field name="theme" type="ThemeConfig" required default="{}"
Theme configuration
:::

::: field name="enabled" type="boolean" optional default="true"
Whether enabled
:::

**Input:**

```md
:::: field-group
::: field name="theme" type="ThemeConfig" required default="{ base: '/' }"
Theme configuration
:::

::: field name="enabled" type="boolean" optional default="true"
Whether enabled
:::

::: field name="callback" type="(...args: any[]) => void" optional default="() => {}"
<Badge type="tip" text="v1.0.0 New"  />
Callback function
:::

::: field name="other" type="string" deprecated
<Badge type="danger" text="v0.9.0 Deprecated"  />
Deprecated property
:::
::::
```

**Output:**

:::: field-group
::: field name="theme" type="ThemeConfig" required default="{ base: '/' }"
Theme configuration
:::

::: field name="enabled" type="boolean" optional default="true"
Whether enabled
:::

::: field name="callback" type="(...args: any[]) => void" optional default="() => {}"
<Badge type="tip" text="v1.0.0 New"  />
Callback function
:::

::: field name="other" type="string" deprecated
<Badge type="danger" text="v0.9.0 Deprecated"  />
Deprecated property
:::
::::
