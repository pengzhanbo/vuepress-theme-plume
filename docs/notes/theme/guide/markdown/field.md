---
title: 字段容器
icon: solar:text-field-linear
createTime: 2025/04/29 09:55:17
permalink: /guide/markdown/field/
badge: 新
---

## 概述

在 markdown 中，使用 `::: field` 容器，用于描述字段信息，包括字段名称、字段类型、是否必填、默认值、详情等信息。

它适用于 描述配置中的字段、组件的 Props 等场景。

还可以使用额外的 `:::: field-group` 容器，用于组合多个 `::: field`。

## 启用

该功能默认不启用，您需要在 `theme` 配置中启用它。

```ts title=".vuepress/config.ts"
export default defineUserConfig({
  theme: plumeTheme({
    markdown: {
      field: true, // [!code ++]
    },
  })
})
```

## 语法

```md
<!-- 单个字段 -->
::: field name="字段名" type="类型" required default="默认值"
字段描述信息
:::

<!-- 字段组合 -->
:::: field-group

::: field name="字段名" type="类型" required default="默认值"
字段描述信息
:::

::: field name="字段名" type="类型" required default="默认值"
字段描述信息
:::

::::
```

## 属性

::: field name="name" required type="string"
字段名称
:::

::: field name="type" type="string" optional
字段类型
:::

::: field name="required" type="boolean" optional
是否必填
:::

::: field name="optional" type="boolean" optional
是否可选
:::

::: field name="default" type="string" optional
默认值
:::

## 示例

**输入：**

```md
::: field name="theme" type="ThemeConfig" required default="{}"
主题配置
:::

::: field name="enabled" type="boolean" optional default="true"
是否启用
:::
```

**输出：**

::: field name="theme" type="ThemeConfig" required default="{}"
主题配置
:::

::: field name="enabled" type="boolean" optional default="true"
是否启用
:::

**输入：**

```md
:::: field-group
::: field name="theme" type="ThemeConfig" required default="{ base: '/' }"
主题配置
:::

::: field name="enabled" type="boolean" optional default="true"
是否启用
:::

::: field name="callback" type="(...args: any[]) => void" optional default="() => {}"
回调函数
:::
::::
```

**输出：**

:::: field-group
::: field name="theme" type="ThemeConfig" required default="{ base: '/' }"
主题配置
:::

::: field name="enabled" type="boolean" optional default="true"
是否启用
:::

::: field name="callback" type="(...args: any[]) => void" optional default="() => {}"
回调函数
:::
::::
