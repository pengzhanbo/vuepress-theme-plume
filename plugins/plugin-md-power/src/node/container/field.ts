import type { Markdown } from 'vuepress/markdown'
import { isUndefined } from '@pengzhanbo/utils'
import { resolveAttrs } from '../utils/resolveAttrs.js'
import { stringifyAttrs } from '../utils/stringifyAttrs.js'
import { createContainerPlugin } from './createContainer.js'

interface FieldAttrs {
  name: string
  type?: string
  required?: boolean
  optional?: boolean
  default?: string
}

export function fieldPlugin(md: Markdown): void {
  createContainerPlugin(md, 'field', {
    before: (info) => {
      const { attrs } = resolveAttrs<FieldAttrs>(info)
      const { name, type, required, optional, default: defaultValue } = attrs
      const props = stringifyAttrs({ name, required, optional })
      return `<VPField${props}${
        !isUndefined(type) ? ` type="${type}"` : ''
      }${
        !isUndefined(defaultValue) ? ` default-value="${defaultValue}"` : ''
      }>`
    },
    after: () => '</VPField>',
  })

  createContainerPlugin(md, 'field-group', {
    before: () => '<div class="vp-field-group">',
  })
}
