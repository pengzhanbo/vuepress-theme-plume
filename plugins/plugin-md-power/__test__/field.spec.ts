import MarkdownIt from 'markdown-it'
import { describe, expect, it } from 'vitest'
import { fieldPlugin } from '../src/node/container/field.js'

describe('fieldPlugin', () => {
  const md = new MarkdownIt().use(fieldPlugin)

  it('::: field', () => {
    const code = `\
::: field name="foo" type="string" required
description
:::

::: field name="bar" type="string" optional
description
:::

::: field name="bar" type="string" deprecated default="baz"
description
:::

::: field name="foo" default="undefined"
description
:::
`
    expect(md.render(code)).toMatchSnapshot()
  })

  it('::: field-group', () => {
    const code = `\
:::: field-group
::: field name="foo" type="string" required
description
:::

::: field name="bar" type="string" optional
description
:::
::::
`
    expect(md.render(code)).toMatchSnapshot()
  })
})
