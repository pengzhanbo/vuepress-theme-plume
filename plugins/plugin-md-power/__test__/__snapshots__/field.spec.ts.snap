// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fieldPlugin > ::: field 1`] = `
"<VPField name="foo" required type="string"><p>description</p>
</VPField><VPField name="bar" optional type="string"><p>description</p>
</VPField><VPField name="bar" deprecated type="string" default-value="baz"><p>description</p>
</VPField><VPField name="foo" default-value="undefined"><p>description</p>
</VPField>"
`;

exports[`fieldPlugin > ::: field-group 1`] = `
"<div class="vp-field-group"><VPField name="foo" required type="string"><p>description</p>
</VPField><VPField name="bar" optional type="string"><p>description</p>
</VPField></div>"
`;
